sql: add infrastructure for unresolved types

rohany opened the pull request:
This PR extends our type system to unresolved/unknown types
before and after type checking. This work is part of the larger project
to add ENUM types to CockroachDB. The idea is to an indirection
to `*types.T` during parsing called a `ResolvableTypeReference`.
Wherever we are unsure whether a type is statically known or not
we have to use this type reference. The typechecking phase then
removes all of these unknown type references by attempting to
resolve each one before proceeding in typechecking. This intuiton
is enforced by ensuring that no values that ascribe to the
`TypedExpr` interface contain `ResolvableTypeReferences`.

A large amount of this commit is propogating this information
through to the rest of the SQL system, where most places expect
`*types.T` we have to teach it to now handle type references and
to resolve them.

Release note: None

rohany committed:
sql: add infrastructure for unresolved types

This PR extends our type system to unresolved/unknown types
before and after type checking. The idea is to an indirection
to `*types.T` during parsing called a `ResolvableTypeReference`.
Wherever we are unsure whether a type is statically known or not
we have to use this type reference. The typechecking phase then
removes all of these unknown type references by attempting to
resolve each one before proceeding in typechecking.

A large amount of this commit is propogating this information
through to the rest of the SQL system, where most places expect
`*types.T` we have to teach it to now handle type references and
to resolve them.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47386)
<!-- Reviewable:end -->

rohany commented:
Now that this is starting to pass tests, I wanted to get opinions on it before I really drill down and fix the remaining tests/removing the tricky TODO's.

cc'ing folks who I think have opinions on the type system -- @knz, @andy-kimball, @RaduBerinde

otan commented on pkg/sql/sem/tree/expr.go:
```diff
@@ -1415,6 +1442,59 @@ const (
 
+// UnresolvedCastExpr represents a CAST(expr AS type) expression with a
+// possibly unresolved type reference.
+type UnresolvedCastExpr struct {
```
do you know why we can't just make `CastExpr` have ResolvableTypeReference?

rohany replied:
There's no technical reason, but it involves changing alot more code -- alot of places make cast exprs, then we have to plumb type resolvers to all those places. Ideally once we typecheck there's no need to resolve anymore types, so that's what I went with here. It should only be before typechecking that if someone has a *types.T they need to wrap it with KnownType before doing anything.

andy-kimball replied:
What about this: We can rename `CastExpr.Type` to `CastExpr.UnresolvedType` (and type it as `ResolvableTypeReference`), and then switch all post-resolution users to call `ResolvedType()` instead. This lets us get rid of `UnresolvedCastExpr` without much disturbance.

otan reviewed (commented):
Took a quick peek but will come back to it - the approach sounds sane and it seems like something postgres has to do something similar to this to resolve types from the name to the entry in `pg_type`. Also similar to the approach of ResolvableFunctionReference if I understand correctly.

The main question I have is whether this affects immutability of the AST. Don't want another ResolvableFunctionReference thing to happen again.

otan commented on pkg/sql/sem/tree/type_name.go:
```diff
@@ -0,0 +1,137 @@
+// in the case that the reference is not statically known. This function
+// is intended to be used in tests or in cases where it is not possible
+// to have any unresolved type references.
+func GetKnownTypeOrPanic(ref ResolvableTypeReference) *types.T {
```
i think the standard is `MustGetKnownType` for things that may panic.

rohany commented:
> The main question I have is whether this affects immutability of the AST. Don't want another ResolvableFunctionReference thing to happen again.

The AST's aren't immutable -- the race was just caused by people assuming it was. But I don't think this affects the immutability -- we aren't writing types into a node once they are resolved.

jordanlewis reviewed (commented):
I feel somewhat unconvinced by this approach so far. I don't see why you need to plumb all of these references so deep into the system. I would prefer an approach that resolved unresolved types into ordinary `types.T` as high up in the stack as possible - but maybe I'm missing some reason this is not possible?

Would it help if you did a pass through the AST before type checking?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M50AhTn0VoMaZ9haeJg:ba4gzfe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/sem/tree/create.go, line 304 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M509gg03PzCAr56OZ4s:-M509gg03PzCAr56OZ4t:b-u59ms5) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/create.go#L304)):*
> ```Go
> 			typ := GetKnownType(d.Type)
> 			if typ == nil {
> 				return nil, errors.New("Cannot collate on this type")
> ```

this should be an internal error
___
*[pkg/sql/sem/tree/eval.go, line 2825 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M509oWh0yyMFMEJMNkI:-M509oWh0yyMFMEJMNkJ:b-jrxcue) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/eval.go#L2825)):*
> ```Go
> 	iVarContainerStack []IndexedVarContainer
> 
> 	TypeResolver TypeReferenceResolver
> ```

comment
___
*[pkg/sql/sem/tree/type_name.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M50A7aZCatWlrn-IKvi:-M50A7aZCatWlrn-IKvj:b-xscfog) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/type_name.go#L40)):*
> ```Go
> // KnownType is a wrapper around *types.T that represents a type reference
> // that has a statically known type.
> type KnownType struct {
> ```

I don't understand the point of this wrapper. Could we have a method on types.T instead? You could add `Resolve` to `types.T` as well, which would do the same thing. I just don't know why you need this new interface set, which seems rather heavy-weight...


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> I would prefer an approach that resolved unresolved types into ordinary types.T as high up in the stack as possible - but maybe I'm missing some reason this is not possible?

It doesn't make sense where to do type resolution before this -- we have a pipeline of parsing -> type checking -> planning -> execution. Typechecking is the first place after parsing where we get a semantic context of how to interpret expressions (i.e. we get a search path etc). This is very similar to the approach to resolve function references.

One reason for the leakage you see is that all of our `Statement`s don't really go through typechecking -- so they hold onto this reference until they try to operate on it later (like add column). This doesn't seem problematic to me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M50CcPMEDqt6-B5MqlV:by23c4q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/sem/tree/type_name.go, line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M50A7aZCatWlrn-IKvi:-M50DQO6ACY9-KfWy3Cp:b-jslg9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/type_name.go#L40)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't understand the point of this wrapper. Could we have a method on types.T instead? You could add `Resolve` to `types.T` as well, which would do the same thing. I just don't know why you need this new interface set, which seems rather heavy-weight...
</blockquote></details>

I agree that its not technically needed -- but there are some logistical issues (not sure how severe they actually are) with pushing this interface down into `types` and letting `types.T` implement it directly:
* This trick of having an unexported method to control what types actually implement the ResolvableType interface won't work (some things that ascribe to the interface in tree won't be able to implement the hidden method)
* This `TypeReferenceResolver` most likely will need to take in a tree.UnresolvedName or something similar in the future, but pushing it down into tree restricts the ability to do that.

Generally the problems seem to be the difficulties caused by types and tree being separate packages. If we let types.T implement this interface directly, we can't add methods on it that use any objects in `tree` in the future, which seems tricky. It would be really nice to have types.T implement this interface directly though -- it would cut a bunch of the changes in this commit out.


<!-- Sent from Reviewable.io -->

otan commented:
> I agree that its not technically needed -- but there are some logistical issues (not sure how severe they actually are) with pushing this interface down into types and letting types.T implement it directly:

currently looking into it, but as a question, what *should* tree contain and what should it not?

otan commented on pkg/cli/dump.go:
```diff
@@ -446,8 +446,13 @@ func getMetadataForTable(conn *sqlConn, md basicMetadata, ts string) (tableMetad
 
+		coltyp := tree.GetKnownType(ref)
+		if coltyp == nil {
+			// TODO (rohany): This should be an unimplemented message.
```
We should be able to do this in this PR I think!

otan reviewed (commented):
if this is the approach we're going with (i don't have any argument against this general approach), it mostly looks good to me (few questions below)

my main gripe/questions are the naming and workflow -- see `type_name.go` comments.

> It would be really nice to have types.T implement this interface directly though -- it would cut a bunch of the changes in this commit out.

is that hard? the interface can still live in tree. i can't make sense of the magnitude of the change atm.

There are also quite a few TODOs in here that seem like we should make them tasks or clarify. I've picked out the ones that I think we should.

andy-kimball reviewed (changes requested):
My initial impression (admittedly not having thought through this problem), is that I think this approach is disruptive and fragile. Any time you have to wrap something as fundamental as a type nearly everywhere it's used, it's a red flag. I think we need to take a step back and re-think this. Do you have any kind of document outlining various approaches you thought through, and why you settled on this one?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M52duea4tZzF3Q_O9FA:b-ea7jqx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)



<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/sql/sem/tree/type_name.go:
```diff
@@ -0,0 +1,138 @@
+// *types.T's. In practice, this will probably be implemented by
+// the planner, but for now it is a dummy interface.
+type TypeReferenceResolver interface {
+	ResolveType() (*types.T, error)
```
I don't get this. The entire catalog is a Resolver, so what does `ResolveType()` return? The type of what? I would expect it to take some argument..

rohany replied:
Sorry, this is still unfinished. I embedded this resolver into different structs to iterate quicker. The end goal is for the `planner` to implement this interface, and pass itself down to the catalog, semaCtx and evalCtx where needed.

I don't know yet what exactly `ResolveType` needs to take in -- once more progress is made on how we are going to perform name resolution of types this will be more clear. As of this PR we don't have any unresolved type names created anywhere yet, so this unused. I want this to look something the corresponding functions on the `tableName` struct that perform name resolution and return descriptors.

rohany commented:
Sorry, i shouldn't have bumped this PR off of being a draft -- there are rough corners as you all have brought up.

I think I'm able to stop wrapping `types.T` and have it implement this interface directly which makes this more palatable and less fragile.

> Do you have any kind of document outlining various approaches you thought through, and why you settled on this one?

Sorry I don't -- I wasn't able to get any alternatives to the point of serious considering. I was hoping that alternatives might arise through discussion here. However, I have feeling that at a high level this is the correct approach, but the actual implementation can be iterated on. The leakage here comes from code structured to use types before/without typechecking, which can't make sense in a world where we allow user defined types.

andy-kimball reviewed (commented):
How do you see a custom type being represented in the type system, once we more fully support them?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M532W_356Du8G1TSMQc:b-wsk6yn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, @RaduBerinde, and @rohany)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
___
*[pkg/sql/opt_catalog.go, line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M545Y-lBsP0JMUPVEFA:-M545Y-lBsP0JMUPVEFB:b-f6p4dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/opt_catalog.go#L40)):*
> ```Go
> // for faster performance.
> type optCatalog struct {
> 	tree.TypeReferenceResolver
> ```

This interface doesn't seem right. Any catalog resolver should return a full schema object, including its `StableID` so that we can detect whether an optimizer dependency has changed. It can't return a simple `types.T` (at least as they stand now) for that reason.

In general, I'd expect that types need to become first-class Schema objects, similar to Tables and Views. IMO, this really needs some kind of RFC or spec, so that we can all get needed context on the high-level proposal before moving forward to implementation.


<!-- Sent from Reviewable.io -->

rohany commented:
> In general, I'd expect that types need to become first-class Schema objects, similar to Tables and Views. IMO, this really needs some kind of RFC or spec, so that we can all get needed context on the high-level proposal before moving forward to implementation.

Discussion for this sort of stuff is going on over in the enums rfc (https://github.com/cockroachdb/cockroach/pull/47070). There's a section in it about type system changes, but I haven't updated it to reflect some choices in this PR.

RaduBerinde reviewed (commented):
I agree it would be useful to propose a catalog API in the RFC. That interface will make it clearer to us what has to happen in the optimizer.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M549A9a5rZ0c9j5reYd:bzftcbt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, @RaduBerinde, and @rohany)



<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/sem/tree/expr.go:
```diff
@@ -1588,7 +1672,49 @@ const (
 
+// UnresolvedAnnotateTypeExpr represents a ANNOTATE_TYPE(expr, type) expression
+// with a possibly unresolved type.
+type UnresolvedAnnotateTypeExpr struct {
```
Same comment here as with `CastExpr`.

andy-kimball commented on pkg/sql/sem/tree/type_name.go:
```diff
@@ -0,0 +1,99 @@
+
+// ArrayTypeReference represents an array of possibly unknown type references.
+type ArrayTypeReference struct {
+	InternalType ResolvableTypeReference
```
Better names: `ElementType` or `ContentType`.

andy-kimball commented:
After reading the RFC, I understand the context of this better. One thing that threw me on first look was the type resolution stuff added to the catalog - it initially appeared like unresolved types were somehow leaking over into a place where they should never be. I wish we didn't have so much code that operates directly on the syntax tree (i.e. AST) rather than the semantic tree (i.e. opt tree), but I guess it is what it is.

In my ideal world, the larger SQL team would come together on a single `Catalog` interface through which all schema-related resolutions occur. This would be based on the `cat.Catalog` interface that the optimizer uses, but expanded to include additional functionality needed by other parts of SQL. We would all cease to directly access fields on the various descriptors, instead using accessor methods. This would give us much better versioning control.

rohany commented:
Hi all, can you PTAL. I've updated this PR heavily according to suggestions and more experience with related parts of the code. I've done the following:

* made *types.T implement ResolvableTypeReference directly (removes alot of the boilerplate)
* hook up the parser so that expressions can be typechecked when provided with a dummy resolver
* Address/remove a majority of the TODO's lying around.
* Done some renames
* Removed the "Unresolved...Expr" pattern and changed uses to call `ResolvedType()` on the expression rather than accessing its type directly.
* Removed attempts at supporting UDT's where they aren't necessary yet
* Refocus this PR to just typechecking of expressions -- we won't support tables of user defined types, so I don't think we need to teach the catalog about this yet. I think that can come in once we start to do name resolution of types.

otan reviewed (commented):
> Reviewed 3 of 43 files at r1, 1 of 14 files at r2, 3 of 33 files at r3, 44 of 50 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, @RaduBerinde, and @rohany)

___
*[pkg/ccl/importccl/read_import_pgdump.go, line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M50QkFs2CTuIVBIU8HR:-M50QkFs2CTuIVBIU8HS:b-5l0ldn) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/ccl/importccl/read_import_pgdump.go#L190)):*
> ```Go
> 					if typ := tree.GetKnownType(e.Type); typ.Oid() == oid.T_regclass {
> 						// tree.Visitor says we should make a copy, but since copyNode is unexported
> 						// and there's no planner here, I think it's safe to directly modify the
> ```

Yikes at the "I think".
___
*[pkg/sql/conn_executor_exec.go, line 332 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iY-11BKgeH0qrL9Ut:-M5iY-12EHtBczqQPTG7:b-xd1mxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/conn_executor_exec.go#L332)):*
> ```Go
> 			typeHints = make(tree.PlaceholderTypes, stmt.NumPlaceholders)
> 			for i, t := range s.Types {
> 				resolved, err := tree.ResolveType(t, ex.planner.semaCtx.GetTypeResolver())
> ```

nit: `typeHints[i], err := ...` works too right?
___
*[pkg/sql/sem/tree/create.go, line 547 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iaxCk5_Ek5C4304BW:-M5iaxCk5_Ek5C4304BX:b-tsorpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/create.go#L547)):*
> ```Go
> 		//  However, the argument is that we deal with serial at parse time only,
> 		//  so we handle those cases here.
> 		switch GetStaticallyKnownType(node.Type).Width() {
> ```

what happens if this is nil?
___
*[pkg/sql/sem/tree/type_check_test.go, line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5ia99TBoMS18GOJcxo:-M5ia99TBoMS18GOJcxp:b-cg85kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_check_test.go#L315)):*
> ```Go
> 	for _, d := range testData {
> 		t.Run(d.expr, func(t *testing.T) {
> 			ctx := tree.MakeSemaContext()
> ```

can we test with explicit nil semaCtx?
___
*[pkg/sql/sem/tree/type_name.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iZRkR4uu211WT2-P6:-M5iZRkR4uu211WT2-P7:b8b1hbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_name.go#L110)):*
> ```Go
> // GetStaticallyKnownType possibly promotes a ResolvableTypeReference into a *types.T
> // if the reference is a statically known type. It returns nil otherwise.
> func GetStaticallyKnownType(ref ResolvableTypeReference) *types.T {
> ```

i know this is painful but...i think to make users think about this properly when using it, it should return `(*types.T, error)`. Wrap the error around something so it's decomposable, e.g. `type NotStaticallyKnownTypeError { error } ` (or something similar).

If that's heavyweight, I think a bool works too, returning false if it is not a statically known type.
___
*[pkg/sql/sem/tree/type_name.go, line 159 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iZhP34SoVUVm4ZOtA:-M5iZhP34SoVUVm4ZOtB:b7euj8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_name.go#L159)):*
> ```Go
> }
> 
> var _ ResolvableTypeReference = &UnresolvedObjectName{}
> ```

nit: i like these closer to the interface definition (or right underneath the struct definition so it's clear `UnresolvedObjectName implements ResolvableTypeReference`).. also (*UnresolvedObjectName)(nil) if you feel poncy.
___
*[pkg/sql/sem/tree/type_name.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iZtJXBwf1PssrULZk:-M5iZtJXBwf1PssrULZl:b-k1snk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_name.go#L178)):*
> ```Go
> 
> // MakeDummyTypeResolver creates a MapTypeResolver from a map.
> func MakeDummyTypeResolver(typeMap map[string]*types.T) TypeReferenceResolver {
> ```

nit: prefix test only with `TestingMapTypeResolver`, and make this name similar.
(also mockery is cool: https://github.com/vektra/mockery, but may be too heavy duty for this)


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/conn_executor_exec.go, line 332 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iY-11BKgeH0qrL9Ut:-M5iuKNZDAI6w4IV7jro:b-mlhl25) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/conn_executor_exec.go#L332)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: `typeHints[i], err := ...` works too right?
</blockquote></details>

oh nvm


<!-- Sent from Reviewable.io -->

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @otan, @RaduBerinde, and @rohany)

___
*[pkg/ccl/importccl/read_import_pgdump.go, line 190 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M50QkFs2CTuIVBIU8HR:-M5j1g339Dd8XkSO7He3:b-7z412e) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/ccl/importccl/read_import_pgdump.go#L190)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yikes at the "I think".
</blockquote></details>

I think this is out of date -- sadly the "i think" was already present, and I don't know enough about this code to prove or disprove it.
___
*[pkg/sql/opt/idxconstraint/index_constraints_test.go, line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvaeIr2sUGOCMznM-r3-89:-M5j5dKiB1U2QgNqBhlT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/opt/idxconstraint/index_constraints_test.go#L89)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think adding support for UDT's is necessary here. We can always add later if we need it.
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/expr_test.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4o66IS8KvnQbanV984-r3-80:-M5j5cv6CsBIyb5XyTVZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/opt/memo/expr_test.go#L80)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think adding support for UDT's is necessary here. We can always add later if we need it.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/builder_test.go, line 84 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4tgYpI152ncU_J13TG-r3-84:-M5j5cidDWZDX6p0DDpY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/opt/optbuilder/builder_test.go#L84)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think adding support for UDT's is necessary here. We can always add later if we need it.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optgen/exprgen/parse_type.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvaeIr2sUGOCMznO-r3-26:-M5j5cKq0JOJ8asbUkPs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/opt/optgen/exprgen/parse_type.go#L26)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think adding support for UDT's is necessary here. We can always add later if we need it. Ditto for all the exprgen-related files.
</blockquote></details>

Done.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 385 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4j3g6aIdLO1utrfPqO-r2-385:-M5j5bwa5y233xVGYcSE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/opt/testutils/testcat/create_table.go#L385)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

maybe confirm with optimizer team and delete the line and replace with an explanation?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/create.go, line 547 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iaxCk5_Ek5C4304BW:-M5j2JF96xj7S_zoOScm:b-61r65h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/create.go#L547)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

what happens if this is nil?
</blockquote></details>

Good catch. Another argument for your suggestion below (or above idk how these files are arranged)
___
*[pkg/sql/sem/tree/eval.go, line 2825 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M509oWh0yyMFMEJMNkI:-M5j5_ON16yspiWDelXo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/eval.go#L2825)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

comment
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/expr.go, line 1447 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvafA9hzExEFtmo_-r1-1447:-M5j5Z9Q6nsbDMabDJzz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/963068e5b9d1089bcb672a4db17ea3a903d3b188/pkg/sql/sem/tree/expr.go#L1447)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What about this: We can rename `CastExpr.Type` to `CastExpr.UnresolvedType` (and type it as `ResolvableTypeReference`), and then switch all post-resolution users to call `ResolvedType()` instead. This lets us get rid of `UnresolvedCastExpr` without much disturbance.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/expr.go, line 1677 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvafA9hzExEFtmo_-r3-1677:-M5j5ZQ4AH_JykZylp3R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/sem/tree/expr.go#L1677)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Same comment here as with `CastExpr`.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_check.go, line 2351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvafA9hzExEFtmoc-r2-2351:-M5j5YMQCtmM0WyxT0oM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sem/tree/type_check.go#L2351)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is this TODO safe to ignore?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_check_test.go, line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5ia99TBoMS18GOJcxo:-M5j5XJN5rEOw8n9pdOW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_check_test.go#L315)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can we test with explicit nil semaCtx?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_name.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvapm0hHz9Eldq7T-r1-72:-M5j5Xq-AUMr7pxAYMVS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/963068e5b9d1089bcb672a4db17ea3a903d3b188/pkg/sql/sem/tree/type_name.go#L72)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

bump!
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_name.go, line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvapm0hHz9Eldq7T-r3-68:-M5j5Syn8Y4HMJo8AirN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4468ab7a3cb56699bb89974c9cf1af2362fb7850/pkg/sql/sem/tree/type_name.go#L68)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Better names: `ElementType` or `ContentType`.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/type_name.go, line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iZRkR4uu211WT2-P6:-M5j2RB0463ovC3rgWXi:b-izrflm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_name.go#L110)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i know this is painful but...i think to make users think about this properly when using it, it should return `(*types.T, error)`. Wrap the error around something so it's decomposable, e.g. `type NotStaticallyKnownTypeError { error } ` (or something similar).

If that's heavyweight, I think a bool works too, returning false if it is not a statically known type.
</blockquote></details>

I think a bool makes sense. Forcing the caller to check the return value will prevent a good number of bugs.
___
*[pkg/sql/sem/tree/type_name.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M5iZtJXBwf1PssrULZk:-M5j5P2BDz2xI797Kubd:b-pbryz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0a3735d586963b80bbff585a4ff765e3257df2f/pkg/sql/sem/tree/type_name.go#L178)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: prefix test only with `TestingMapTypeResolver`, and make this name similar.
(also mockery is cool: https://github.com/vektra/mockery, but may be too heavy duty for this)
</blockquote></details>

pulling in dependencies is a nono
___
*[pkg/sql/sqlbase/testutils.go, line 1321 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M4fWvapm0hHz9Eldq7X-r2-1321:-M5j5gZr7xLJFMC7LUs1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10850404482aec87c1fc01b5180282528ed6a33d/pkg/sql/sqlbase/testutils.go#L1321)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

will we?
(what's our policy on cleaning up TODOs? should this be TODO(task num) instead?)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan reviewed (approved):
changes LGTM!

> Reviewed 9 of 19 files at r5, 1 of 12 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M5voBA5504yDyAeHazZ:b-17k651)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @otan, @RaduBerinde, and @rohany)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
Thanks for all the work on this. It's an ambitious PR, with a lot of details to get right. Looks like you addressed all of my concerns. I didn't look at every detail, but the overall structure and design looks right to me.

rohany commented:
TFTRs!

bors r+

jordanlewis reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> this is great, exciting! I have a few tiny nits, but don't bother un-borsing it since I know this is going to take more work to finish soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-:-M6CnfhvChhHtHb8PXNg:b-st03sh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @otan, @RaduBerinde, and @rohany)

___
*[pkg/sql/sem/tree/name_part.go, line 211 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M6Clfdn2C5Lpy05IB56:-M6Clfdo5xb4ih3pkM5A:btsifft) ([raw file](https://github.com/cockroachdb/cockroach/blob/30dd78002f5dee009f70dfab104893dd7263e9bd/pkg/sql/sem/tree/name_part.go#L211)):*
> ```Go
> func (u *UnresolvedName) ToUnresolvedObjectName(idx AnnotationIdx) (*UnresolvedObjectName, error) {
> 	if u.NumParts == 4 {
> 		return nil, errors.Newf("improper qualified name (too many dotted names): %s", u)
> ```

Is this a user-facing error? If so, it should have a pgcode.
___
*[pkg/sql/sem/tree/type_name.go, line 71 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M6Clqhe0x0xFwLbE3-W:-M6Clqhe0x0xFwLbE3-X:bswsxo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/30dd78002f5dee009f70dfab104893dd7263e9bd/pkg/sql/sem/tree/type_name.go#L71)):*
> ```Go
> 
> // TypeReferenceResolver is the interface that will provide the ability
> // to actually lookup type metadata and transform references into
> ```

nit: s/lookup/look up/
___
*[pkg/sql/sem/tree/type_name.go, line 104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M6Cm9v78zrR_XJucOHr:-M6Cm9v78zrR_XJucOHs:bx8w3vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/30dd78002f5dee009f70dfab104893dd7263e9bd/pkg/sql/sem/tree/type_name.go#L104)):*
> ```Go
> 			// If we don't have a resolver, we can't actually resolve this
> 			// name into a type.
> 			return nil, errors.Newf("type %q does not exist", t)
> ```

This probably wants to have a pgcode as well.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/sem/tree/type_name.go, line 128 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47386#-M6CzJS-8lpiAynCEMIF:-M6CzJS-8lpiAynCEMIG:b-2f5jgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/30dd78002f5dee009f70dfab104893dd7263e9bd/pkg/sql/sem/tree/type_name.go#L128)):*
> ```Go
> 		return typ
> 	}
> 	panic("type reference was not a statically known type")
> ```

[nit] panic(errors.AssertionFailedf(..))


<!-- Sent from Reviewable.io -->

rohany commented:
I'll just fix these, bors seems unhappy.

bors r-

rohany commented:
bors r+
